File: app/lib/domain/entities/booking.dart
import 'package:equatable/equatable.dart';
import 'flight.dart';

enum BookingStatus {
  created,
  confirmed,
  cancelled,
}

class Ticket extends Equatable {
  final int id;
  final String ticketNumber;
  final String seatNumber;

  const Ticket({
    required this.id,
    required this.ticketNumber,
    required this.seatNumber,
  });

  @override
  List<Object?> get props => [id, ticketNumber, seatNumber];
}

class Booking extends Equatable {
  final int id;
  final String pnr;
  final int userId;
  final int flightId;
  final BookingStatus status;
  final DateTime? seatHoldExpiresAt;
  final DateTime createdAt;
  final Flight? flight;
  final List<Ticket> tickets;
  final String? passengerName;

  const Booking({
    required this.id,
    required this.pnr,
    required this.userId,
    required this.flightId,
    required this.status,
    this.seatHoldExpiresAt,
    required this.createdAt,
    this.flight,
    required this.tickets,
    this.passengerName,
  });

  bool get isHoldExpired =>
      seatHoldExpiresAt != null && DateTime.now().isAfter(seatHoldExpiresAt!);

  @override
  List<Object?> get props => [
        id,
        pnr,
        userId,
        flightId,
        status,
        seatHoldExpiresAt,
        createdAt,
        flight,
        tickets,
        passengerName,
      ];
}

File: AirlineAIT/backend/app/models/booking.py
"""
Booking, Ticket, and SeatHold models.

Booking: Ties user to flight with PNR, status (HOLD, CONFIRMED, CANCELLED, EXPIRED).
Ticket: Individual passenger tickets within a booking.
SeatHold: Temporary seat reservations during booking process.

Part of: Backend Models
"""
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Enum, Float, Boolean
from sqlalchemy.orm import relationship
import enum
from typing import Optional
from app.db.base import Base


class BookingStatus(str, enum.Enum):
    HOLD = "HOLD"
    CONFIRMED = "CONFIRMED"
    CANCELLED = "CANCELLED"
    EXPIRED = "EXPIRED"


class Booking(Base):
    __tablename__ = "bookings"

    id = Column(Integer, primary_key=True, index=True)
    pnr = Column(String, unique=True, nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    flight_id = Column(Integer, ForeignKey("flights.id"), nullable=False)
    status = Column(Enum(BookingStatus), default=BookingStatus.HOLD, nullable=False)
    created_at = Column(DateTime, nullable=False)
    
    # Relationships
    user = relationship("User", back_populates="bookings")
    flight = relationship("Flight", back_populates="bookings")
    tickets = relationship("Ticket", back_populates="booking", cascade="all, delete-orphan")
    payments = relationship("Payment", back_populates="booking", cascade="all, delete-orphan")


class Ticket(Base):
    __tablename__ = "tickets"

    id = Column(Integer, primary_key=True, index=True)
    ticket_number = Column(String, unique=True, nullable=False, index=True)
    booking_id = Column(Integer, ForeignKey("bookings.id"), nullable=False)
    passenger_profile_id = Column(Integer, ForeignKey("passenger_profiles.id"), nullable=False)
    seat_number = Column(String, nullable=True)  # e.g., "12A"
    price = Column(Float, nullable=True)  # Fixed price at booking time
    
    # Relationships
    booking = relationship("Booking", back_populates="tickets")
    passenger = relationship("PassengerProfile", back_populates="tickets")
    check_in = relationship("CheckIn", back_populates="ticket", uselist=False, cascade="all, delete-orphan")

    @property
    def passenger_name(self) -> Optional[str]:
        return self.passenger.full_name if self.passenger else None

    @property
    def is_checked_in(self) -> bool:
        return self.check_in is not None


class SeatHold(Base):
    __tablename__ = "seat_holds"

    id = Column(Integer, primary_key=True, index=True)
    flight_id = Column(Integer, ForeignKey("flights.id"), nullable=False)
    seat_number = Column(String, nullable=False)
    held_until = Column(DateTime, nullable=False)
    booking_id = Column(Integer, ForeignKey("bookings.id"), nullable=True)
    
    # Relationships
    flight = relationship("Flight", back_populates="seat_holds")
    booking = relationship("Booking", backref="seat_holds")

